### PR TITLE
Radmap#3

### PR DESCRIPTION
**Pull request by tripledoublev**
Refactoring:
- Created `common.rs`  
- Moved constants to their own module.  
- Moved tests to `lib.rs`.  

Implemented features:
- Created CommonKeypair struct and DHTEntity trait in `common.rs` file.
- Added create_repo and get_repo functions.

Enhancements:
- Updated Group struct to include repos and implement the DHTEntity trait.
- Updated Repo struct with necessary fields and implemented the DHTEntity trait.
- Replaced CryptoKey with SharedSecret for `encryption_key`.
- Ensured `encryption_key` uses `random_shared_secret`.

### CODE REVIEW

File: src/backend.rs [hunk 1]
```diff
@@ -1,18 +1,23 @@
 use std::path::{Path, PathBuf};
 use std::collections::HashMap;
-use eyre::{Result, anyhow};
+use anyhow::{Result, anyhow};
```
Comment (RangerMauve) [inline src/backend.rs]: Why use anyhow instead?

File: src/backend.rs [hunk 2]
```diff
@@ -1,18 +1,23 @@
...
-use crate::group::{Group, GroupKeypair};
+use crate::common::{CommonKeypair, DHTEntity};
+use crate::group::Group;
+use crate::repo::Repo;
 
 pub struct Backend {
     path: PathBuf,
     port: u16,
-    veilid_api: Option<VeilidAPI>,
+    veilid_api: Option<Arc<VeilidAPI>>,
```
Comment (RangerMauve) [inline src/backend.rs]: Does this have to be an Arc? Maybe we could use a Box instead?

File: src/backend.rs [hunk 3]
```diff
@@ -60,15 +68,23 @@ impl Backend {
...
+        }
 
         Ok(())
     }
 
     pub async fn stop(&self) -> Result<()> {
         println!("Stopping Backend...");
         if let Some(veilid) = &self.veilid_api {
-            veilid.clone().shutdown().await;
+            if let Ok(veilid) = Arc::try_unwrap(Arc::clone(veilid)) {
```
Comment (RangerMauve) [inline src/backend.rs]: What is the purpose of doing a clone? Can we not shutdown the instance held in the arc?

**Review by RangerMauve — CHANGES_REQUESTED** [review]

File: src/backend.rs (hunk 1 above)
Comment (tripledoublev) [reply src/backend.rs]: The switch from eyre to anyhow is to ensure compatibility with the ? operator.
anyhow's error type implements std::error::Error and can be used in contexts where ? operator is applied

File: src/backend.rs (hunk 2 above)
Comment (tripledoublev) [reply src/backend.rs]: If I switch to a Box, we find ourselves faced with a lack of persistence after shutdown.
It gets back to the initial issue where with the Box, the shutdown is successful but I can not retrieve the group name after shutdown

Comment (RangerMauve) [reply src/backend.rs]: I think that's okay. shutdown should be the last thing called and afterwards you should need to start veilid again to get a group and call get name on it.

Comment (RangerMauve) [reply src/backend.rs]: We can have the group id cloned to reference the group in between shutdowns

Comment (RangerMauve) [reply src/backend.rs]: If the table store isn't persisting group info, or the dht isn't persisting data we published, then we've got more of a problem.

File: src/group.rs [hunk 4]
```diff
@@ -1,134 +1,88 @@
...
-    pub secret_key: Option<CryptoKey>,
-    pub encryption_key: CryptoKey,
-}
+use crate::common::DHTEntity;
+use crate::repo::Repo;
 
 #[derive(Clone)]
 pub struct Group {
     pub id: CryptoKey,
+    pub record_key: TypedKey,
```
Comment (RangerMauve) [inline src/group.rs]: The DHTRecordDescriptor has a lot of this data already, I think we should do the following:

- remove `id/record_key/secret_key` from the Group struct
- have `id()` be a getter for `dht_Record.key`
- add a getter for `owner_key` and `owner_secret`
- Refactor the code to reference the group by the record key and pull the keypair out just for init
- the encryption key should stay the same

**Review by RangerMauve — CHANGES_REQUESTED** [review]
Looks great, good job! Just one thing for cleaning up the group struct a little

File: src/group.rs (hunk 4 above)
Comment (tripledoublev) [reply src/group.rs]: Thank you for such clear guidelines! 

I’ve made the following updates:

- Removed id, record_key, and secret_key from the Group struct.
- Added id(), owner_key(), and owner_secret() as getters from dht_record.
- Refactored the code to reference the group by the record key and only pull the keypair for initialization.
- Kept the encryption key unchanged as requested.

That really helped cleaning it up.
Thanks again!

File: src/backend.rs [hunk 5]
```diff
@@ -139,33 +139,29 @@ impl Backend {
...
-            record_key,
             dht_record,
             encryption_key,
-            Some(CryptoTyped::new(CRYPTO_KIND_VLD0, keypair.secret)),
             Arc::new(routing_context),
             crypto_system,
         );
     
         let protected_store = veilid.protected_store().unwrap();
         CommonKeypair {
```
Comment (RangerMauve) [inline src/backend.rs:155]: This struct should have the id under a new `id` field and the public_key should be the owner public key.

**Review by RangerMauve — CHANGES_REQUESTED** [review]
Are the tests passing still?

File: src/backend.rs (hunk 5 above)
Comment (tripledoublev) [reply src/backend.rs:155]: Just to be sure which struct we are talking about. Are you saying I should add a new `id` field to the group struct?

Comment (RangerMauve) [reply src/backend.rs:155]: To the commonkeypair struct

Comment (tripledoublev) [reply src/backend.rs:155]: - Included the id field to the keypair struct

**Comment by tripledoublev** [comment]
> Are the tests passing still?

Tests are passing

**Review by RangerMauve — APPROVED** [review]
Looks great! 🥳

**Pull request merged**